Use the node identifier instead of the node path

remuslazar opened the pull request:
Use the node identifier instead of the node path

Using the node path for identifying the node in ES causes some issues while moving the node to another path. This commit changes this behavior, generating an unique ES identifier for the record using the node identifier instead of using the node path.

remuslazar committed:
Use the node identifier instead of the node path

Using the node path for identifying the node in ES causes some issues while moving the node to another path. This commit changes this behavior, generating an unique ES identifier for the record using the node identifier instead of using the node path.

remuslazar committed:
don't hash the contextIdentifier

for easier debugging, don't hash the contextIdentifier, use it directly for the Elasticsearch _id. So instead of having _id's like e.g. bd6dabcf6c7b2c48c48d3ee11434f99581ed12d4, the _id's will be like 416733dc-b040-5147-3407-5727ebce8fe5-live.

baschny commented on Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/NodeIndexer.php:
```diff
@@ -110,6 +110,16 @@ public function getIndex() {
+	 * @return string
+	 */
+	protected function getContextIdentifier(NodeInterface $node) {
+		return $node->getIdentifier() . '-' . $node->getWorkspace()->getName();
```
nicht besser ein "@" als separator, damit das zumindest im look&feel wie bei dem Trenner zwischen Pfad und Workspace ist?

remuslazar replied:
war auch meine Idee, geht aber nicht, das erlaubt ES nicht

baschny replied:
und wenn man es "genau" nimmt, macht das Originale (getContextPath) noch die Dimension dazu:

```
    if ($this->context->getDimensions() !== array()) {
        $contextPath .= ';';
        foreach ($this->context->getDimensions() as $dimensionName => $dimensionValues) {
            $contextPath .= $dimensionName . '=' . implode(',', $dimensionValues) . '&';
        }
        $contextPath = substr($contextPath, 0, -1);
    }
```

kann man vermutlich auch gleich mit einbauen, damit das auch unterstützt wird.

baschny replied:
Und das "@" sollte schon gehen laut ES und z.B. http://stackoverflow.com/questions/12168260/elasticsearch-id-constraints gibt es da keine Begrenzungen. Muss man halt vielleicht encoden wenn man das via REST aufruft, evtl ist das das "Problem" (weil das nciht richtig gemacht wird). Kann man also so lassen.

baschny commented on Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/NodeIndexer.php:
```diff
@@ -367,7 +375,7 @@ public function removeNode(NodeInterface $node) {
 		);
 
-		$this->logger->log(sprintf('NodeIndexer: Removed node %s from index (node actually removed). Persistence ID: %s', $node->getContextPath(), $identifier), LOG_DEBUG, NULL, 'ElasticSearch (CR)');
+		$this->logger->log(sprintf('NodeIndexer: Removed %s from index (node actually removed). Persistence ID: %s', $node, $identifier), LOG_DEBUG, NULL, 'ElasticSearch (CR)');
```
war im Original schon falsch, das letzte in dem String ist doch nicht der "Persistence ID" sondern der "Identifier". Das sind doch separate Sachen oder? Persistence ID ist doch die ID in der Datenbank, nach einem Reimport eine andere, "Identifier" ist das globale immer gültige was auch site export / import überlebt. Oder??

remuslazar committed:
getContextIdentifier() refactoring

use the existing getContextPath() method to build our ContextIdentifier. Basically just replacing the path there (everything before the @ sign) by the node identifier, to be more future-proof and DRY.

remuslazar committed:
cleanups in "Removed %s from index" log message

The Log-Message was not clear about the meaning of the persistence ID.

remuslazar commented:
@baschny habe nun beide Sachen refactored, schau mal!

kdambekalns commented on Classes/Flowpack/ElasticSearch/ContentRepositoryAdaptor/Indexer/NodeIndexer.php:
```diff
@@ -110,6 +110,18 @@ public function getIndex() {
+	protected function getContextIdentifier(NodeInterface $node) {
+		$contextPath = $node->getContextPath();
+
+		return $node->getIdentifier() . substr($contextPath, strpos($contextPath, '@'));
```
The use of substr() should be replaced by using getWorkspace()->getName() instead, analogous to getContextPath().

baschny replied:
Solved in e7d0ce4079848ab7a0bcb20c6e879ed89b7453ee

baschny commented:
Will rebase on psr-2
